src/connector/src/parser/postgres.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chrono::{NaiveDate, Utc};
+use risingwave_common::catalog::Schema;
+use risingwave_common::row::OwnedRow;
+use risingwave_common::types::{
+    DataType, Date, Decimal, Interval, JsonbVal, ListValue, ScalarImpl, Time, Timestamp,
+    Timestamptz,
+};
+use rust_decimal::Decimal as RustDecimal;
+
+use crate::source::cdc::external::ConnectorResult;
+
+macro_rules! handle_data_type {
+    ($row:expr, $i:expr, $type:ty, $builder:expr) => {
+        let v = $row.try_get::<_, Option<Vec<$type>>>($i)?;
+        v.map(|vec| {
+            vec.into_iter()
+                .for_each(|val| $builder.append(Some(ScalarImpl::from(val))))
+        });
+    };
+    ($row:expr, $i:expr, $type:ty, $builder:expr, $rw_type:ty) => {
+        let v = $row.try_get::<_, Option<Vec<$type>>>($i)?;
+        v.map(|vec| {
+            vec.into_iter()
+                .for_each(|val| $builder.append(Some(ScalarImpl::from(<$rw_type>::from(val)))))
+        });
+    };
+}
+
+pub fn postgres_row_to_owned_row(
+    row: tokio_postgres::Row,
+    schema: &Schema,
+) -> ConnectorResult<OwnedRow> {
+    let mut datums = vec![];
+    for i in 0..schema.fields.len() {
+        let rw_field = &schema.fields[i];
+        let datum = {
+            match &rw_field.data_type {
+                DataType::Boolean => {
+                    let v = row.try_get::<_, Option<bool>>(i)?;
+                    v.map(ScalarImpl::from)
+                }
+                DataType::Int16 => {
+                    let v = row.try_get::<_, Option<i16>>(i)?;
+                    v.map(ScalarImpl::from)
+                }
+                DataType::Int32 => {
+                    let v = row.try_get::<_, Option<i32>>(i)?;
+                    v.map(ScalarImpl::from)
+                }
+                DataType::Int64 => {
+                    let v = row.try_get::<_, Option<i64>>(i)?;
+                    v.map(ScalarImpl::from)
+                }
+                DataType::Float32 => {
+                    let v = row.try_get::<_, Option<f32>>(i)?;
+                    v.map(ScalarImpl::from)
+                }
+                DataType::Float64 => {
+                    let v = row.try_get::<_, Option<f64>>(i)?;
+                    v.map(ScalarImpl::from)
+                }
+                DataType::Decimal => {
+                    let v = row.try_get::<_, Option<RustDecimal>>(i)?;
```

Comment:
test with `'nan'::decimal`?

Reply:
'nan'::decimal cannot be correctly parsed. https://github.com/risingwavelabs/risingwave/issues/14218